File: Combination_Sum.cpp
class Solution {
public:
    /**
     * @param candidates: A list of integers
     * @param target:An integer
     * @return: A list of lists of integers
     */
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        // write your code here
        vector<vector<int>> ret;
        vector<int> solution;
        sort(candidates.begin(), candidates.end());
        helper(0, candidates, target, solution, ret);
        return ret;
    }
private:
    void helper(int start, vector<int> &candidates, int left, vector<int> &solution, vector<vector<int>> &ret)
    {
        if (left == 0){
            ret.emplace_back(solution);
            return;
        }
        else if (left < 0)
        {
            return;
        }
        for (int i = start; i < candidates.size(); ++i)
        {
            if (i > 0 && candidates[i] == candidates[i - 1])
            {
                continue;
            }
            solution.emplace_back(candidates[i]);
            helper(i, candidates, left - candidates[i], solution, ret);
            solution.pop_back();
        }
    }
};

File: Combination Sum.py
class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum(self, candidates, target):
        # write your code here
        def helper(start, left, tmp, ret, candidates, length):
            if left == 0:
                ret.append(tmp[:])
                return
            else:
                for i in range(start, length):
                    if candidates[i] > left:
                        break
                    else:
                        tmp.append(candidates[i])
                        helper(i, left - candidates[i], tmp, ret, candidates, length)
                        tmp.pop()
        if not candidates:
            return []
        candidates.sort()
        tmp, ret = [], []
        helper(0, target, tmp, ret, candidates, len(candidates))
        
        return ret